### PR TITLE
feat(24.04): Add slice definitions for libgif7, libpng16-16t64 and metapixel

### PR DESCRIPTION
**Pull request by sjauld**
# Proposed changes

Adds slices for libgif7, libpng16-16t64 and metapixel

GIFLIB is a package of portable tools and library routines for working with GIF images.

libpng is a library implementing an interface for reading and writing PNG (Portable Network Graphics) format files.

Metapixel is a program for generating photomosaics.

## Related issues/PRs
N/A

### Forward porting
N/A

## Checklist

* [x] I have read the [contributing guidelines](
https://github.com/canonical/chisel-releases/blob/main/CONTRIBUTING.md)
* [x] I have tested my changes ([see how](https://github.com/canonical/chisel-releases/blob/main/CONTRIBUTING.md#7-test-your-slices-before-opening-a-pr))
* [x] I have already submitted the [CLA form](
https://ubuntu.com/legal/contributors/agreement)

### CODE REVIEW

**Review by rebornplusplus — COMMENTED** [review]
Looks good to me, thank you!

Can we please add any spread tests (in `/tests/spread/integration/libgif7`) for these slices? It could be something basic, like opening a gif file or writing one or something else.

Also note that the CLA check is failing, which means you need to sign the CLA. Let me know if you already have, and the check is failing nonetheless!

**Comment by rebornplusplus** [comment]
> I wasn't able to find an example of how to test a library using spread.

Just noticed this. Is there any utility/binary which uses the `libgif7` libs to open/write/create a gif image? If so, we can check with that.

**Comment by sjauld** [comment]
Hi @rebornplusplus,

I did sign the CLA before submitting.

I'll have another look at the tests, although I struggled to find any examples.

Cheers,
Stuart

**Comment by sjauld** [comment]
> > I wasn't able to find an example of how to test a library using spread.
> 
> Just noticed this. Is there any utility/binary which uses the `libgif7` libs to open/write/create a gif image? If so, we can check with that.

Seems likely. I'll see what I can do

**Comment by sjauld** [comment]
We can test with giftext, which is part of the ubuntu package giflib-tools, using the below command which will fail if the library isn't available.

But I'm not sure how to get giftext into the spread environment... any ideas?

```
echo "\x47\x49\x46\x38\x37\x61\x1\x0\x1\x0\x80\x1\x0\x0\x0\x0\x4d\x21\x35\x2c\x0\x0\x0\x0\x1\x0\x1\x0\x0\x2\x2\x4c\x1\x0\x3b" | giftext
```

**Comment by rebornplusplus** [comment]
> We can test with giftext, which is part of the ubuntu package giflib-tools, using the below command which will fail if the library isn't available.
> 
> But I'm not sure how to get giftext into the spread environment... any ideas?
> 
> ```
> echo "\x47\x49\x46\x38\x37\x61\x1\x0\x1\x0\x80\x1\x0\x0\x0\x0\x4d\x21\x35\x2c\x0\x0\x0\x0\x1\x0\x1\x0\x0\x2\x2\x4c\x1\x0\x3b" | giftext
> ```

We can probably install (download and extract I think) the binaries from `giflib-tools` in the chroot folder and then test there. Would probably be a bit complext really. It would have been much easier if the `giflib-tools` package was chiselled as well. 

I guess we could only test if the appropriate library file `libgif.so.7` is present or not. That should be something like this: 

https://github.com/canonical/chisel-releases/blob/4373954f8f6ae9ff4e5fa763798e838eaff3e59c/tests/spread/integration/base-passwd/task.yaml#L6

**Comment by sjauld** [comment]
The problem with testing the location of the library is it's in a different spot for each architecture, and I couldn't work out how to manage that in spread.

I might just chisel giflib-tools as well, that should be pretty straight forward and allow for a more realistic test.

**Comment by sjauld** [comment]
Ok giflib-tools depends on perl... so I have chiselled metapixel instead and written tests.

## libgif7 test run

```sh
2024-08-01 10:56:15 Project content is packed for delivery (489.79KB).
2024-08-01 10:56:15 If killed, discard servers with: spread -reuse-pid=3864072 -discard
2024-08-01 10:56:15 Allocating docker:ubuntu-24.04-s390x...
2024-08-01 10:56:15 Allocating docker:ubuntu-24.04-amd64...
2024-08-01 10:56:15 Allocating docker:ubuntu-24.04-arm64v8...
2024-08-01 10:56:15 Allocating docker:ubuntu-24.04-arm32v7...
2024-08-01 10:56:15 Allocating docker:ubuntu-24.04-ppc64le...
2024-08-01 10:56:46 Waiting for docker:ubuntu-24.04-amd64 to make SSH available at 10.10.0.6...
2024-08-01 10:56:46 Allocated docker:ubuntu-24.04-amd64.
2024-08-01 10:56:46 Connecting to docker:ubuntu-24.04-amd64...
2024-08-01 10:56:46 Connected to docker:ubuntu-24.04-amd64 at 10.10.0.6.
2024-08-01 10:56:46 Sending project content to docker:ubuntu-24.04-amd64...
2024-08-01 10:56:46 Preparing docker:ubuntu-24.04-amd64 (docker:ubuntu-24.04-amd64)...
2024-08-01 10:56:54 Preparing docker:ubuntu-24.04-amd64:tests/spread/integration/libgif7 (docker:ubuntu-24.04-amd64)...
2024-08-01 10:56:54 Executing docker:ubuntu-24.04-amd64:tests/spread/integration/libgif7 (docker:ubuntu-24.04-amd64) (1/5)...
2024-08-01 10:57:08 Discarding docker:ubuntu-24.04-amd64...
2024-08-01 11:01:46 Waiting for docker:ubuntu-24.04-s390x to make SSH available at 10.10.0.2...
2024-08-01 11:01:46 Allocated docker:ubuntu-24.04-s390x.
2024-08-01 11:01:46 Connecting to docker:ubuntu-24.04-s390x...
2024-08-01 11:01:46 Connected to docker:ubuntu-24.04-s390x at 10.10.0.2.
2024-08-01 11:01:46 Sending project content to docker:ubuntu-24.04-s390x...
2024-08-01 11:01:46 Preparing docker:ubuntu-24.04-s390x (docker:ubuntu-24.04-s390x)...
2024-08-01 11:02:07 Preparing docker:ubuntu-24.04-s390x:tests/spread/integration/libgif7 (docker:ubuntu-24.04-s390x)...
2024-08-01 11:02:07 Executing docker:ubuntu-24.04-s390x:tests/spread/integration/libgif7 (docker:ubuntu-24.04-s390x) (2/5)...
2024-08-01 11:02:31 Waiting for docker:ubuntu-24.04-arm32v7 to make SSH available at 10.10.0.4...
2024-08-01 11:02:31 Allocated docker:ubuntu-24.04-arm32v7.
2024-08-01 11:02:31 Connecting to docker:ubuntu-24.04-arm32v7...
2024-08-01 11:02:32 Connected to docker:ubuntu-24.04-arm32v7 at 10.10.0.4.
2024-08-01 11:02:32 Sending project content to docker:ubuntu-24.04-arm32v7...
2024-08-01 11:02:32 Preparing docker:ubuntu-24.04-arm32v7 (docker:ubuntu-24.04-arm32v7)...
2024-08-01 11:02:40 Discarding docker:ubuntu-24.04-s390x...
2024-08-01 11:02:53 Preparing docker:ubuntu-24.04-arm32v7:tests/spread/integration/libgif7 (docker:ubuntu-24.04-arm32v7)...
2024-08-01 11:02:54 Executing docker:ubuntu-24.04-arm32v7:tests/spread/integration/libgif7 (docker:ubuntu-24.04-arm32v7) (3/5)...
2024-08-01 11:02:57 Waiting for docker:ubuntu-24.04-ppc64le to make SSH available at 10.10.0.5...
2024-08-01 11:02:57 Allocated docker:ubuntu-24.04-ppc64le.
2024-08-01 11:02:57 Connecting to docker:ubuntu-24.04-ppc64le...
2024-08-01 11:02:57 Connected to docker:ubuntu-24.04-ppc64le at 10.10.0.5.
2024-08-01 11:02:57 Sending project content to docker:ubuntu-24.04-ppc64le...
2024-08-01 11:02:58 Preparing docker:ubuntu-24.04-ppc64le (docker:ubuntu-24.04-ppc64le)...
2024-08-01 11:03:18 Preparing docker:ubuntu-24.04-ppc64le:tests/spread/integration/libgif7 (docker:ubuntu-24.04-ppc64le)...
2024-08-01 11:03:18 Executing docker:ubuntu-24.04-ppc64le:tests/spread/integration/libgif7 (docker:ubuntu-24.04-ppc64le) (4/5)...
2024-08-01 11:04:56 Discarding docker:ubuntu-24.04-arm32v7...
2024-08-01 11:05:05 Discarding docker:ubuntu-24.04-ppc64le...
2024-08-01 11:08:17 Waiting for docker:ubuntu-24.04-arm64v8 to make SSH available at 10.10.0.3...
2024-08-01 11:08:17 Allocated docker:ubuntu-24.04-arm64v8.
2024-08-01 11:08:17 Connecting to docker:ubuntu-24.04-arm64v8...
2024-08-01 11:08:17 Connected to docker:ubuntu-24.04-arm64v8 at 10.10.0.3.
2024-08-01 11:08:17 Sending project content to docker:ubuntu-24.04-arm64v8...
2024-08-01 11:08:18 Preparing docker:ubuntu-24.04-arm64v8 (docker:ubuntu-24.04-arm64v8)...
2024-08-01 11:09:25 Preparing docker:ubuntu-24.04-arm64v8:tests/spread/integration/libgif7 (docker:ubuntu-24.04-arm64v8)...
2024-08-01 11:09:25 Executing docker:ubuntu-24.04-arm64v8:tests/spread/integration/libgif7 (docker:ubuntu-24.04-arm64v8) (5/5)...
2024-08-01 11:12:02 Discarding docker:ubuntu-24.04-arm64v8...
2024-08-01 11:12:03 Successful tasks: 5
2024-08-01 11:12:03 Aborted tasks: 0
```

## libpng16-16t64 test run

```sh
2024-08-01 11:12:56 Project content is packed for delivery (490.33KB).
2024-08-01 11:12:56 If killed, discard servers with: spread -reuse-pid=3944686 -discard
2024-08-01 11:12:56 Allocating docker:ubuntu-24.04-ppc64le...
2024-08-01 11:12:56 Allocating docker:ubuntu-24.04-s390x...
2024-08-01 11:12:56 Allocating docker:ubuntu-24.04-amd64...
2024-08-01 11:12:56 Allocating docker:ubuntu-24.04-arm64v8...
2024-08-01 11:12:56 Allocating docker:ubuntu-24.04-arm32v7...
2024-08-01 11:13:32 Waiting for docker:ubuntu-24.04-amd64 to make SSH available at 10.10.0.4...
2024-08-01 11:13:32 Allocated docker:ubuntu-24.04-amd64.
2024-08-01 11:13:32 Connecting to docker:ubuntu-24.04-amd64...
2024-08-01 11:13:32 Connected to docker:ubuntu-24.04-amd64 at 10.10.0.4.
2024-08-01 11:13:32 Sending project content to docker:ubuntu-24.04-amd64...
2024-08-01 11:13:32 Preparing docker:ubuntu-24.04-amd64 (docker:ubuntu-24.04-amd64)...
2024-08-01 11:13:39 Preparing docker:ubuntu-24.04-amd64:tests/spread/integration/libpng16-16t64 (docker:ubuntu-24.04-amd64)...
2024-08-01 11:13:39 Executing docker:ubuntu-24.04-amd64:tests/spread/integration/libpng16-16t64 (docker:ubuntu-24.04-amd64) (1/5)...
2024-08-01 11:13:52 Discarding docker:ubuntu-24.04-amd64...
2024-08-01 11:18:26 Waiting for docker:ubuntu-24.04-s390x to make SSH available at 10.10.0.3...
2024-08-01 11:18:26 Allocated docker:ubuntu-24.04-s390x.
2024-08-01 11:18:26 Connecting to docker:ubuntu-24.04-s390x...
2024-08-01 11:18:27 Connected to docker:ubuntu-24.04-s390x at 10.10.0.3.
2024-08-01 11:18:27 Sending project content to docker:ubuntu-24.04-s390x...
2024-08-01 11:18:27 Preparing docker:ubuntu-24.04-s390x (docker:ubuntu-24.04-s390x)...
2024-08-01 11:18:42 Waiting for docker:ubuntu-24.04-arm32v7 to make SSH available at 10.10.0.2...
2024-08-01 11:18:42 Allocated docker:ubuntu-24.04-arm32v7.
2024-08-01 11:18:42 Connecting to docker:ubuntu-24.04-arm32v7...
2024-08-01 11:18:43 Connected to docker:ubuntu-24.04-arm32v7 at 10.10.0.2.
2024-08-01 11:18:43 Sending project content to docker:ubuntu-24.04-arm32v7...
2024-08-01 11:18:43 Preparing docker:ubuntu-24.04-arm32v7 (docker:ubuntu-24.04-arm32v7)...
2024-08-01 11:18:47 Preparing docker:ubuntu-24.04-s390x:tests/spread/integration/libpng16-16t64 (docker:ubuntu-24.04-s390x)...
2024-08-01 11:18:47 Executing docker:ubuntu-24.04-s390x:tests/spread/integration/libpng16-16t64 (docker:ubuntu-24.04-s390x) (2/5)...
2024-08-01 11:19:03 Preparing docker:ubuntu-24.04-arm32v7:tests/spread/integration/libpng16-16t64 (docker:ubuntu-24.04-arm32v7)...
2024-08-01 11:19:03 Executing docker:ubuntu-24.04-arm32v7:tests/spread/integration/libpng16-16t64 (docker:ubuntu-24.04-arm32v7) (3/5)...
2024-08-01 11:19:08 Waiting for docker:ubuntu-24.04-ppc64le to make SSH available at 10.10.0.6...
2024-08-01 11:19:08 Allocated docker:ubuntu-24.04-ppc64le.
2024-08-01 11:19:08 Connecting to docker:ubuntu-24.04-ppc64le...
2024-08-01 11:19:08 Connected to docker:ubuntu-24.04-ppc64le at 10.10.0.6.
2024-08-01 11:19:08 Sending project content to docker:ubuntu-24.04-ppc64le...
2024-08-01 11:19:08 Preparing docker:ubuntu-24.04-ppc64le (docker:ubuntu-24.04-ppc64le)...
2024-08-01 11:19:29 Preparing docker:ubuntu-24.04-ppc64le:tests/spread/integration/libpng16-16t64 (docker:ubuntu-24.04-ppc64le)...
2024-08-01 11:19:29 Executing docker:ubuntu-24.04-ppc64le:tests/spread/integration/libpng16-16t64 (docker:ubuntu-24.04-ppc64le) (4/5)...
2024-08-01 11:20:49 Discarding docker:ubuntu-24.04-s390x...
2024-08-01 11:21:12 Discarding docker:ubuntu-24.04-arm32v7...
2024-08-01 11:21:35 Discarding docker:ubuntu-24.04-ppc64le...
2024-08-01 11:23:57 Waiting for docker:ubuntu-24.04-arm64v8 to make SSH available at 10.10.0.5...
2024-08-01 11:23:57 Allocated docker:ubuntu-24.04-arm64v8.
2024-08-01 11:23:57 Connecting to docker:ubuntu-24.04-arm64v8...
2024-08-01 11:23:57 Connected to docker:ubuntu-24.04-arm64v8 at 10.10.0.5.
2024-08-01 11:23:57 Sending project content to docker:ubuntu-24.04-arm64v8...
2024-08-01 11:23:58 Preparing docker:ubuntu-24.04-arm64v8 (docker:ubuntu-24.04-arm64v8)...
2024-08-01 11:25:00 Preparing docker:ubuntu-24.04-arm64v8:tests/spread/integration/libpng16-16t64 (docker:ubuntu-24.04-arm64v8)...
2024-08-01 11:25:00 Executing docker:ubuntu-24.04-arm64v8:tests/spread/integration/libpng16-16t64 (docker:ubuntu-24.04-arm64v8) (5/5)...
2024-08-01 11:27:44 Discarding docker:ubuntu-24.04-arm64v8...
2024-08-01 11:27:45 Successful tasks: 5
2024-08-01 11:27:45 Aborted tasks: 0
```

## metapixel test run

```sh
2024-08-01 11:30:39 Project content is packed for delivery (490.33KB).
2024-08-01 11:30:39 If killed, discard servers with: spread -reuse-pid=4036232 -discard
2024-08-01 11:30:39 Allocating docker:ubuntu-24.04-arm32v7...
2024-08-01 11:30:39 Allocating docker:ubuntu-24.04-s390x...
2024-08-01 11:30:39 Allocating docker:ubuntu-24.04-ppc64le...
2024-08-01 11:30:39 Allocating docker:ubuntu-24.04-arm64v8...
2024-08-01 11:30:39 Allocating docker:ubuntu-24.04-amd64...
2024-08-01 11:31:10 Waiting for docker:ubuntu-24.04-amd64 to make SSH available at 10.10.0.5...
2024-08-01 11:31:10 Allocated docker:ubuntu-24.04-amd64.
2024-08-01 11:31:10 Connecting to docker:ubuntu-24.04-amd64...
2024-08-01 11:31:10 Connected to docker:ubuntu-24.04-amd64 at 10.10.0.5.
2024-08-01 11:31:10 Sending project content to docker:ubuntu-24.04-amd64...
2024-08-01 11:31:10 Preparing docker:ubuntu-24.04-amd64 (docker:ubuntu-24.04-amd64)...
2024-08-01 11:31:17 Preparing docker:ubuntu-24.04-amd64:tests/spread/integration/metapixel (docker:ubuntu-24.04-amd64)...
2024-08-01 11:31:17 Executing docker:ubuntu-24.04-amd64:tests/spread/integration/metapixel (docker:ubuntu-24.04-amd64) (1/5)...
2024-08-01 11:31:27 Discarding docker:ubuntu-24.04-amd64...
2024-08-01 11:35:45 Waiting for docker:ubuntu-24.04-ppc64le to make SSH available at 10.10.0.6...
2024-08-01 11:35:45 Allocated docker:ubuntu-24.04-ppc64le.
2024-08-01 11:35:45 Connecting to docker:ubuntu-24.04-ppc64le...
2024-08-01 11:35:45 Connected to docker:ubuntu-24.04-ppc64le at 10.10.0.6.
2024-08-01 11:35:45 Sending project content to docker:ubuntu-24.04-ppc64le...
2024-08-01 11:35:46 Preparing docker:ubuntu-24.04-ppc64le (docker:ubuntu-24.04-ppc64le)...
2024-08-01 11:35:54 Waiting for docker:ubuntu-24.04-arm32v7 to make SSH available at 10.10.0.2...
2024-08-01 11:35:54 Allocated docker:ubuntu-24.04-arm32v7.
2024-08-01 11:35:54 Connecting to docker:ubuntu-24.04-arm32v7...
2024-08-01 11:35:55 Connected to docker:ubuntu-24.04-arm32v7 at 10.10.0.2.
2024-08-01 11:35:55 Sending project content to docker:ubuntu-24.04-arm32v7...
2024-08-01 11:35:55 Preparing docker:ubuntu-24.04-arm32v7 (docker:ubuntu-24.04-arm32v7)...
2024-08-01 11:36:06 Preparing docker:ubuntu-24.04-ppc64le:tests/spread/integration/metapixel (docker:ubuntu-24.04-ppc64le)...
2024-08-01 11:36:06 Executing docker:ubuntu-24.04-ppc64le:tests/spread/integration/metapixel (docker:ubuntu-24.04-ppc64le) (2/5)...
2024-08-01 11:36:14 Preparing docker:ubuntu-24.04-arm32v7:tests/spread/integration/metapixel (docker:ubuntu-24.04-arm32v7)...
2024-08-01 11:36:15 Executing docker:ubuntu-24.04-arm32v7:tests/spread/integration/metapixel (docker:ubuntu-24.04-arm32v7) (3/5)...
2024-08-01 11:36:25 Waiting for docker:ubuntu-24.04-s390x to make SSH available at 10.10.0.4...
2024-08-01 11:36:25 Allocated docker:ubuntu-24.04-s390x.
2024-08-01 11:36:25 Connecting to docker:ubuntu-24.04-s390x...
2024-08-01 11:36:25 Connected to docker:ubuntu-24.04-s390x at 10.10.0.4.
2024-08-01 11:36:25 Sending project content to docker:ubuntu-24.04-s390x...
2024-08-01 11:36:25 Preparing docker:ubuntu-24.04-s390x (docker:ubuntu-24.04-s390x)...
2024-08-01 11:36:43 Preparing docker:ubuntu-24.04-s390x:tests/spread/integration/metapixel (docker:ubuntu-24.04-s390x)...
2024-08-01 11:36:43 Executing docker:ubuntu-24.04-s390x:tests/spread/integration/metapixel (docker:ubuntu-24.04-s390x) (4/5)...
2024-08-01 11:38:00 Discarding docker:ubuntu-24.04-ppc64le...
2024-08-01 11:38:13 Discarding docker:ubuntu-24.04-arm32v7...
2024-08-01 11:38:33 Discarding docker:ubuntu-24.04-s390x...
2024-08-01 11:40:28 Waiting for docker:ubuntu-24.04-arm64v8 to make SSH available at 10.10.0.3...
2024-08-01 11:40:28 Allocated docker:ubuntu-24.04-arm64v8.
2024-08-01 11:40:28 Connecting to docker:ubuntu-24.04-arm64v8...
2024-08-01 11:40:29 Connected to docker:ubuntu-24.04-arm64v8 at 10.10.0.3.
2024-08-01 11:40:29 Sending project content to docker:ubuntu-24.04-arm64v8...
2024-08-01 11:40:29 Preparing docker:ubuntu-24.04-arm64v8 (docker:ubuntu-24.04-arm64v8)...
2024-08-01 11:41:31 Preparing docker:ubuntu-24.04-arm64v8:tests/spread/integration/metapixel (docker:ubuntu-24.04-arm64v8)...
2024-08-01 11:41:31 Executing docker:ubuntu-24.04-arm64v8:tests/spread/integration/metapixel (docker:ubuntu-24.04-arm64v8) (5/5)...
2024-08-01 11:43:31 Discarding docker:ubuntu-24.04-arm64v8...
2024-08-01 11:43:31 Successful tasks: 5
2024-08-01 11:43:31 Aborted tasks: 0
```

**Comment by cjdcordeiro** [comment]
> Diff of dependencies:
> 
> slices/metapixel.yaml
> ```diff
> @@ -1,4 +1,4 @@
>  libc6
>  libgif7
> -libjpeg8
> +libjpeg-turbo8
>  libpng16-16t64
> ```

It also looks good to me. I'd just like to have some clarifications on this diff here

**Comment by rebornplusplus** [comment]
> > Diff of dependencies:
> > slices/metapixel.yaml
> > ```diff
> > @@ -1,4 +1,4 @@
> >  libc6
> >  libgif7
> > -libjpeg8
> > +libjpeg-turbo8
> >  libpng16-16t64
> > ```
> 
> It also looks good to me. I'd just like to have some clarifications on this diff here

I knew this looked familiar. Had a similar issue at https://github.com/canonical/chisel-releases/pull/254#discussion_r1696769672.

**Comment by sjauld** [comment]
I have changed that reference back to libjpeg8 (and added the libjpeg8 wrapper slice)

**Review by cjdcordeiro — APPROVED** [review]
thanks LGTM

File: slices/libgif7.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# library for GIF images
```
Comment (rebornplusplus) [inline slices/libgif7.yaml:1]:
```suggestion
# Library for GIF images
```

File: slices/metapixel.yaml [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# generator for photomosaics
```
Comment (rebornplusplus) [inline slices/metapixel.yaml:1]:
```suggestion
# Metapixel is a program for generating photomosaics.
```

**Review by rebornplusplus — APPROVED** [review]
Looks good to me, thank you!

Only left a few nitpicks.

**Review by rebornplusplus — APPROVED** [review]
Looks good to me, thank you!

**Pull request merged**